server/sonar-server/src/main/java/org/sonar/server/plugins/PluginDownloader.java

```diff
@@ -20,35 +20,150 @@
     this.updateCenterMatrixFactory = updateCenterMatrixFactory;
+    this.downloader = downloader;
+    this.downloadDir = fileSystem.getDownloadedPluginsDir();
+  }
+
+  /**
+   * Deletes the temporary files remaining from previous downloads
+   */
+  @Override
+  public void start() {
+    try {
+      forceMkdir(downloadDir);
+      for (File tempFile : listTempFile(this.downloadDir)) {
+        deleteQuietly(tempFile);
+      }
+    } catch (IOException e) {
+      throw new IllegalStateException("Fail to create the directory: " + downloadDir, e);
+    }
+  }
+
+  @Override
+  public void stop() {
+    // Nothing to do
+  }
+
+  public void cancelDownloads() {
+    try {
+      if (downloadDir.exists()) {
+        org.sonar.core.util.FileUtils.cleanDirectory(downloadDir);
+      }
+    } catch (IOException e) {
+      throw new IllegalStateException("Fail to clean the plugin downloads directory: " + downloadDir, e);
+    }
+  }
+
+  public boolean hasDownloads() {
+    return !getDownloadedPluginFilenames().isEmpty();
+  }
+
+  public List<String> getDownloadedPluginFilenames() {
+    List<String> names = new ArrayList<>();
+    for (File file : listPlugins(this.downloadDir)) {
+      names.add(file.getName());
+    }
+    return names;
+  }
+
+  /**
+   * @return the list of download plugins as {@link PluginInfo} instances
+   */
+  public Collection<PluginInfo> getDownloadedPlugins() {
+    return listPlugins(this.downloadDir)
+      .stream()
+      .map(PluginInfo::create)
+      .collect(MoreCollectors.toList());
   }
 
   public void download(String pluginKey, Version version) {
     Optional<UpdateCenter> updateCenter = updateCenterMatrixFactory.getUpdateCenter(true);
     if (updateCenter.isPresent()) {
       List<Release> installablePlugins = updateCenter.get().findInstallablePlugins(pluginKey, version);
       checkRequest(!installablePlugins.isEmpty(), "Error while downloading plugin '%s' with version '%s'. No compatible plugin found.", pluginKey, version.getName());
-      for (Release r : installablePlugins) {
-        super.download(r);
+      for (Release release : installablePlugins) {
+        try {
+          downloadRelease(release);
+        } catch (Exception e) {
+          String message = String.format("Fail to download the plugin (%s, version %s) from %s (error is : %s)",
+            release.getArtifact().getKey(), release.getVersion().getName(), release.getDownloadUrl(), e.getMessage());
+          LOG.debug(message, e);
+          throw new IllegalStateException(message, e);
+        }
+      }
+    }
+  }
+
+  private void downloadRelease(Release release) throws URISyntaxException, IOException {
+    String url = release.getDownloadUrl();
+
+    URI uri = new URI(url);
+    if (url.startsWith("file:")) {
+      // used for tests
+      File file = toFile(uri.toURL());
+      copyFileToDirectory(file, downloadDir);
+    } else {
+      String filename = substringAfterLast(uri.getPath(), "/");
```

Comment:
lines 150 to 123 ought to be shared across the 2 download implementation.
This is support for updatecenter inconsistencies or for some "old" plugins. This is not download logic per say.